<a href="https://colab.research.google.com/github/Lorenzito/ProgramacionParaAnaliticaDescriptivayPredicitva/blob/main/Actividad_Calificada_07_Datos_Duplicados_273524.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Actividad calificada: Datos duplicados**
### Detección, cuantificación y tratamiento de duplicados en un dataset real

---

| | |
|---|---|
| **Materia** | Programación para Analítica Descriptiva y Predictiva |
| **Programa** | Maestría en Inteligencia Artificial y Analítica de Datos (MIAAD) — UACJ |
| **Unidad** | 02 Análisis Descriptivo de los Datos |
| **Tema** | Limpieza y transformación de datos: duplicados |
| **Modalidad** | Individual |
| **Valor** | 100 puntos |
| **Dataset** | Hotel Booking Demand (Antonio, Almeida y Nunes, 2019) |
| **Fecha de entrega** | 25/Sep/26 |

---

## Datos del estudiante

| | |
|---|---|
| **Nombre completo** | Lorenzo Varela Ollervides |
| **Matrícula** | 273524 |

## Instrucciones

1. Trabaja en Google Colab. Guarda una copia en tu Drive (**Archivo → Guardar una copia en Drive**) antes de empezar.
2. Completa tus datos en la tabla anterior.
3. Resuelve las actividades **en orden**: algunas usan variables creadas en actividades anteriores.
4. Escribe tu código en las celdas marcadas con `# Escribe tu código aquí`. Puedes agregar celdas si lo necesitas.
5. **Comenta tu código**: cada bloque debe explicar qué hace y por qué. Los comentarios forman parte de la calificación de cada actividad.
6. Responde las preguntas en las celdas **Respuesta**, con tus propias palabras y apoyándote en los resultados que obtuviste (cita cifras).
7. Usa los métodos vistos en el notebook de la sesión: `duplicated()`, `drop_duplicates()`, `value_counts()`, `sort_values()`, `groupby()`, `pd.concat()`, máscaras booleanas, etc.
8. Antes de entregar, ejecuta todo el notebook (**Entorno de ejecución → Reiniciar y ejecutar todo**) y verifica que no haya errores.
9. **Entrega en Moodle** un archivo de texto `Actividad_Duplicados_<Matricula>.txt` que contenga el **enlace** a tu notebook (Google Colab o GitHub), con las salidas visibles.
   - Colab: **Compartir → Acceso general → Cualquier persona con el enlace (Lector)**.
   - GitHub: el repositorio debe ser **público**.
   - Opcional: puedes adjuntar también el archivo `.ipynb`.

## Contexto

El dataset **Hotel Booking Demand** contiene 119,390 reservas de dos hoteles de Portugal (un hotel urbano, *City Hotel*, y un hotel vacacional, *Resort Hotel*) con llegada entre julio de 2015 y agosto de 2017. Cada fila es una reserva y hay 32 columnas. Algunas relevantes para esta actividad:

| Columna | Descripción |
|---|---|
| `hotel` | `City Hotel` o `Resort Hotel` |
| `is_canceled` | 1 si la reserva se canceló, 0 si no |
| `lead_time` | Días entre la reserva y la fecha de llegada |
| `arrival_date_year`, `arrival_date_month`, `arrival_date_day_of_month` | Fecha de llegada |
| `adults`, `children`, `babies` | Número de huéspedes |
| `country` | País de origen (código ISO de 3 letras) |
| `market_segment` | Segmento de mercado (`Online TA`, `Groups`, `Direct`, …) |
| `deposit_type` | Tipo de depósito (`No Deposit`, `Non Refund`, `Refundable`) |
| `agent`, `company` | Identificador de la agencia o empresa que hizo la reserva (con faltantes) |
| `customer_type` | Tipo de cliente (`Transient`, `Group`, …) |
| `adr` | Tarifa diaria promedio (*average daily rate*) |
| `reservation_status` | Estado final: `Check-Out`, `Canceled`, `No-Show` |
| `reservation_status_date` | Fecha del último cambio de estado |

> **Importante**: el dataset **no incluye un identificador de reserva**. Tenlo presente al interpretar los duplicados.

## Rúbrica

| Actividad | Puntos | Excelente | Bueno | Insuficiente | No entregado / Incorrecto |
|---|---|---|---|---|---|
| **1. Diagnóstico inicial** | 20 | 18–20: conteos y porcentajes correctos con los tres valores de `keep`; duplicados por hotel calculados; código comentado; conclusión con cifras. | 14–17: resultados correctos con omisiones menores en comentarios o en la conclusión. | 1–13: conteos incompletos o con errores; conclusión sin sustento. | 0: no se entregó o el código no produce resultados. |
| **2. Inspección y criterio** | 25 | 23–25: grupos inspeccionados y ordenados; tamaños de grupo analizados; comparación duplicados vs. total; argumento sólido sobre si son errores, con evidencia. | 18–22: análisis correcto pero con argumento parcial o poca evidencia. | 1–17: inspección superficial; argumento sin datos. | 0: no se entregó o es incorrecto. |
| **3. Impacto en el análisis** | 25 | 23–25: métricas por hotel antes y después con `groupby()`; tabla comparativa con `pd.concat()`; interpretación del efecto en el análisis y en un modelo predictivo. | 18–22: métricas correctas; tabla o interpretación incompletas. | 1–17: métricas con errores o sin comparación. | 0: no se entregó o es incorrecto. |
| **4. Tratamiento y verificación** | 20 | 18–20: limpieza verificada; duplicados parciales detectados e interpretados; registro de decisión completo. | 14–17: limpieza correcta; análisis parcial o registro incompleto. | 1–13: limpieza sin verificación o análisis parcial erróneo. | 0: no se entregó o es incorrecto. |
| **Reflexión final** | 10 | 9–10: respuestas fundamentadas en los resultados propios. | 7–8: respuestas correctas pero generales. | 1–6: respuestas vagas o sin relación con los resultados. | 0: no se entregó. |
| **Total** | **100** | | | | |

**Notas**:

- Las respuestas escritas de cada actividad (1.4, 2.4, 3.4, 4.2 y 4.3) y los comentarios del código cuentan para la calificación **de esa actividad**, no solo para la reflexión final.
- Entrega tardía: se descuenta **5 % por día natural** de retraso, hasta 7 días; después de ese plazo la calificación es 0. El descuento se aplica sobre la calificación obtenida.

## Carga de datos

Ejecuta la siguiente celda. No requiere modificaciones.

In [1]:
import pandas as pd

# Carga del dataset Hotel Booking Demand desde GitHub (TidyTuesday, 2020-02-11)
url = 'https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-02-11/hotels.csv'
hoteles = pd.read_csv(url)

# Dimensiones y primeras filas
print('Dimensiones:', hoteles.shape)
hoteles.head()

Dimensiones: (119390, 32)


,hotel,is_canceled,lead_time,arrival_date_year,arrival_date_month,arrival_date_week_number,arrival_date_day_of_month,stays_in_weekend_nights,stays_in_week_nights,adults,...,deposit_type,agent,company,days_in_waiting_list,customer_type,adr,required_car_parking_spaces,total_of_special_requests,reservation_status,reservation_status_date
0,Resort Hotel,0,342,2015,July,27,1,0,0,2,...,No Deposit,NaN,NaN,0,Transient,0.0,0,0,Check-Out,2015-07-01
1,Resort Hotel,0,737,2015,July,27,1,0,0,2,...,No Deposit,NaN,NaN,0,Transient,0.0,0,0,Check-Out,2015-07-01
2,Resort Hotel,0,7,2015,July,27,1,0,1,1,...,No Deposit,NaN,NaN,0,Transient,75.0,0,0,Check-Out,2015-07-02
3,Resort Hotel,0,13,2015,July,27,1,0,1,1,...,No Deposit,304.0,NaN,0,Transient,75.0,0,0,Check-Out,2015-07-02
4,Resort Hotel,0,14,2015,July,27,1,0,2,2,...,No Deposit,240.0,NaN,0,Transient,98.0,0,1,Check-Out,2015-07-03


---
## Actividad 1. Diagnóstico inicial (20 puntos)

### 1.1 Conteo de duplicados exactos (5 pts)

Calcula el número de filas marcadas como duplicadas en `hoteles` considerando **todas** las columnas, con `keep='first'`, `keep='last'` y `keep=False`. Imprime los tres resultados con un texto que explique qué significa cada uno.

In [2]:
# duplicated() compara TODAS las columnas (subset=None): detecta duplicados exactos.
# .sum() cuenta los True (True = 1, False = 0).

# keep='first': marca todas las repeticiones excepto la primera aparición de cada grupo
n_first = hoteles.duplicated(keep='first').sum()

# keep='last': marca todas las repeticiones excepto la última aparición de cada grupo
n_last = hoteles.duplicated(keep='last').sum()

# keep=False: marca TODAS las filas que pertenecen a un grupo repetido, incluida la que se conservaría
n_todas = hoteles.duplicated(keep=False).sum()

print(f"keep='first': {n_first:,} filas sobrantes (copias de una fila que ya apareció antes)")
print(f"keep='last' : {n_last:,} filas sobrantes (copias de una fila que vuelve a aparecer después)")
print(f"keep=False  : {n_todas:,} filas involucradas en duplicados (todas las filas de cada grupo repetido)")

keep='first': 31,994 filas sobrantes (copias de una fila que ya apareció antes)
keep='last' : 31,994 filas sobrantes (copias de una fila que vuelve a aparecer después)
keep=False  : 40,165 filas involucradas en duplicados (todas las filas de cada grupo repetido)


### 1.2 Magnitud del problema (5 pts)

Con los resultados de 1.1, calcula e imprime:

- El **porcentaje** de filas sobrantes respecto al total.
- El número de **grupos repetidos** (combinaciones de valores que aparecen más de una vez).

In [3]:
# Porcentaje de filas sobrantes: la media de la máscara booleana es la proporción de True
porcentaje_sobrantes = hoteles.duplicated().mean() * 100

# Grupos repetidos: filas involucradas - filas sobrantes = una fila conservada por cada grupo
n_grupos = n_todas - n_first

print(f'Total de filas         : {len(hoteles):,}')
print(f'Filas sobrantes        : {n_first:,}')
print(f'Porcentaje de sobrantes: {porcentaje_sobrantes:.2f} %')
print(f'Grupos repetidos       : {n_grupos:,}')

Total de filas         : 119,390
Filas sobrantes        : 31,994
Porcentaje de sobrantes: 26.80 %
Grupos repetidos       : 8,171


### 1.3 Duplicados por hotel (5 pts)

Determina **cuántas** filas sobrantes (`keep='first'`) hay en cada hotel y qué **porcentaje** representan respecto al total de reservas de ese hotel.

Sugerencia: crea una copia de `hoteles`, agrega una columna booleana `es_duplicado` con el resultado de `duplicated()` y agrupa por `hotel`.

In [4]:
# Copia para no modificar el DataFrame original
hoteles_dup = hoteles.copy()

# Columna booleana: True si la fila es una copia sobrante (keep='first')
hoteles_dup['es_duplicado'] = hoteles.duplicated(keep='first')

# Por hotel: size() da el total de reservas, sum() cuenta los True y mean() da su proporción
duplicados_hotel = pd.DataFrame({
    'reservas':   hoteles_dup.groupby('hotel').size(),
    'sobrantes':  hoteles_dup.groupby('hotel')['es_duplicado'].sum(),
    'porcentaje': hoteles_dup.groupby('hotel')['es_duplicado'].mean() * 100
})

print(duplicados_hotel)

              reservas  sobrantes  porcentaje
hotel                                        
City Hotel       79330      25902   32.650952
Resort Hotel     40060       6092   15.207189


### 1.4 Conclusión (5 pts)

¿Qué tan grave es el problema de duplicados en este dataset? ¿Afecta por igual a ambos hoteles? Responde con cifras.

**Respuesta 1.4:**

El problema es **grave**. De 119,390 filas, 31,994 son copias exactas de otra fila (**26.80 %**), es decir, más de una de cada cuatro reservas está repetida en las 32 columnas. En total, 40,165 filas pertenecen a alguno de los **8,171 grupos repetidos**. Los tres valores de `keep` son coherentes: `'first'` y `'last'` dan el mismo número (31,994), porque solo cambian *cuáles* filas se marcan; `keep=False` da 40,165 = 31,994 + 8,171.

**No afecta por igual a ambos hoteles**. El *City Hotel* tiene 25,902 filas sobrantes, el **32.65 %** de sus 79,330 reservas, mientras que el *Resort Hotel* tiene 6,092, el **15.21 %** de sus 40,060. La proporción de duplicados en el hotel urbano es más del doble que en el vacacional, así que cualquier comparación entre hoteles hecha con los datos originales está sesgada.

---
## Actividad 2. Inspección y criterio de duplicado (25 puntos)

### 2.1 Ver los grupos repetidos (6 pts)

Filtra las filas involucradas en duplicados (`keep=False`), **ordénalas** con `sort_values()` para que las copias queden juntas y muestra las primeras 12 filas con estas columnas:

`hotel`, `lead_time`, `arrival_date_year`, `arrival_date_month`, `arrival_date_day_of_month`, `adults`, `country`, `market_segment`, `deposit_type`, `adr`, `reservation_status`.

Elige tú las columnas de ordenamiento y explica tu elección en un comentario.

In [5]:
# Máscara con TODAS las filas que forman parte de un grupo repetido
mascara_dup = hoteles.duplicated(keep=False)

# Columnas que se mostrarán
columnas_vista = ['hotel', 'lead_time', 'arrival_date_year', 'arrival_date_month',
                  'arrival_date_day_of_month', 'adults', 'country', 'market_segment',
                  'deposit_type', 'adr', 'reservation_status']

# Criterio de ordenamiento: primero las columnas de la vista (hotel, anticipación, fecha de
# llegada, etc.), para que la tabla se lea fácil, y después el RESTO de las columnas.
# Como los duplicados son exactos (coinciden en las 32 columnas), ordenar por todas ellas
# garantiza que las copias de un mismo grupo queden en filas consecutivas; si ordenáramos
# solo por algunas, dos grupos que empatan en esas columnas podrían quedar intercalados.
columnas_orden = columnas_vista + list(hoteles.columns.drop(columnas_vista))

# Filtramos, ordenamos y mostramos las primeras 12 filas con las columnas pedidas
hoteles[mascara_dup].sort_values(by=columnas_orden)[columnas_vista].head(12)

,hotel,lead_time,arrival_date_year,arrival_date_month,arrival_date_day_of_month,adults,country,market_segment,deposit_type,adr,reservation_status
40643,City Hotel,0,2015,August,4,2,PRT,Direct,No Deposit,0.0,Canceled
40644,City Hotel,0,2015,August,4,2,PRT,Direct,No Deposit,0.0,Canceled
40772,City Hotel,0,2015,August,7,2,PRT,Direct,No Deposit,75.0,Check-Out
40802,City Hotel,0,2015,August,7,2,PRT,Direct,No Deposit,75.0,Check-Out
40821,City Hotel,0,2015,August,8,2,PRT,Online TA,No Deposit,89.0,Check-Out
40838,City Hotel,0,2015,August,8,2,PRT,Online TA,No Deposit,89.0,Check-Out
76792,City Hotel,0,2015,August,10,2,FRA,Online TA,No Deposit,75.0,Check-Out
76793,City Hotel,0,2015,August,10,2,FRA,Online TA,No Deposit,75.0,Check-Out
76794,City Hotel,0,2015,August,10,2,FRA,Online TA,No Deposit,75.0,Check-Out
41067,City Hotel,0,2015,August,11,1,PRT,Corporate,No Deposit,88.0,Check-Out


### 2.2 Tamaño de los grupos (6 pts)

Usa `value_counts()` sobre el `DataFrame` completo para contar cuántas veces aparece cada combinación de filas. Considera que las columnas `agent` y `company` tienen faltantes (revisa el parámetro `dropna`).

Con ese resultado:

1. Muestra **cuántos grupos** hay de cada tamaño (2 copias, 3 copias, …). Basta con los 10 tamaños más frecuentes.
2. Indica el **tamaño del grupo más grande** y muestra su combinación de valores.

In [6]:
# value_counts() sobre el DataFrame completo: cuenta cuántas veces aparece cada fila (combinación).
# dropna=False es necesario porque agent y company tienen NaN: con el valor por defecto
# (dropna=True) se descartan las combinaciones que contienen algún NaN.
tamanos = hoteles.value_counts(dropna=False)

# Comprobación del efecto de dropna: filas que quedan contadas con cada opción
print(f'Filas contadas con dropna=False: {tamanos.sum():,}')
print(f'Filas contadas con dropna=True : {hoteles.value_counts().sum():,}')

# 1) Solo grupos repetidos (aparecen más de una vez) y cuántos grupos hay de cada tamaño
print('\nNúmero de grupos según su tamaño (10 tamaños más frecuentes):')
print(tamanos[tamanos > 1].value_counts().head(10))

# 2) Tamaño del grupo más grande (value_counts ordena de mayor a menor)
print(f'\nTamaño del grupo más grande: {tamanos.max()} copias')

Filas contadas con dropna=False: 119,390
Filas contadas con dropna=True : 217

Número de grupos según su tamaño (10 tamaños más frecuentes):
count
2     5034
3      969
4      497
5      275
6      171
10     123
7      115
8      103
20      97
9       93
Name: count, dtype: int64

Tamaño del grupo más grande: 180 copias


In [7]:
# Combinación de valores del grupo más grande: value_counts() ordena de mayor a menor,
# así que es el primer elemento del índice de tamanos (una tupla con un valor por columna).
# zip() empareja cada nombre de columna con su valor para mostrar las 32 columnas completas
print(f'Combinación que se repite {tamanos.max()} veces:\n')
for columna, valor in zip(hoteles.columns, tamanos.index[0]):
    print(f'{columna:32}: {valor}')

Combinación que se repite 180 veces:

hotel                           : City Hotel
is_canceled                     : 1
lead_time                       : 277
arrival_date_year               : 2016
arrival_date_month              : November
arrival_date_week_number        : 46
arrival_date_day_of_month       : 7
stays_in_weekend_nights         : 1
stays_in_week_nights            : 2
adults                          : 2
children                        : 0.0
babies                          : 0
meal                            : BB
country                         : PRT
market_segment                  : Groups
distribution_channel            : TA/TO
is_repeated_guest               : 0
previous_cancellations          : 0
previous_bookings_not_canceled  : 0
reserved_room_type              : A
assigned_room_type              : A
booking_changes                 : 0
deposit_type                    : Non Refund
agent                           : nan
company                         : nan
days_in_waiti

### 2.3 ¿En qué se distinguen las filas duplicadas? (6 pts)

Compara las filas involucradas en duplicados (`keep=False`) con el dataset completo en:

- Distribución de `market_segment` (proporciones).
- Distribución de `deposit_type` (proporciones).
- Tasa de cancelación (`is_canceled`).

In [8]:
# Filas involucradas en duplicados (misma máscara de 2.1, keep=False)
duplicadas = hoteles[mascara_dup]

# value_counts(normalize=True) devuelve proporciones; se colocan lado a lado en un DataFrame
# (las Series se alinean por su índice, es decir, por la categoría)
print('Distribución de market_segment (proporciones):')
print(pd.DataFrame({
    'duplicadas': duplicadas['market_segment'].value_counts(normalize=True),
    'total':      hoteles['market_segment'].value_counts(normalize=True)
}))

print('\nDistribución de deposit_type (proporciones):')
print(pd.DataFrame({
    'duplicadas': duplicadas['deposit_type'].value_counts(normalize=True),
    'total':      hoteles['deposit_type'].value_counts(normalize=True)
}))

# La media de is_canceled (0/1) es la tasa de cancelación
print(f"\nTasa de cancelación en filas duplicadas: {duplicadas['is_canceled'].mean():.4f}")
print(f"Tasa de cancelación en el total        : {hoteles['is_canceled'].mean():.4f}")

Distribución de market_segment (proporciones):
                duplicadas     total
market_segment                      
Aviation          0.000473  0.001985
Complementary     0.001867  0.006223
Corporate         0.036699  0.044350
Direct            0.034981  0.105587
Groups            0.416109  0.165935
Offline TA/TO     0.303996  0.202856
Online TA         0.205876  0.473046
Undefined              NaN  0.000017

Distribución de deposit_type (proporciones):
              duplicadas     total
deposit_type                      
No Deposit      0.641678  0.876464
Non Refund      0.356430  0.122179
Refundable      0.001892  0.001357

Tasa de cancelación en filas duplicadas: 0.5839
Tasa de cancelación en el total        : 0.3704


### 2.4 Criterio (7 pts)

El dataset no tiene identificador de reserva. Con base en la evidencia de 2.1, 2.2 y 2.3, responde:

a) ¿Las filas duplicadas son **errores de registro** o podrían ser **reservas distintas** que coinciden en todos sus valores? Argumenta con datos.

b) Las columnas `agent` y `company` tienen muchos faltantes. ¿Cómo trata `duplicated()` esos `NaN` y qué implicación tiene para este análisis?

**Respuesta 2.4:**

**a)** La evidencia indica que **muchas de estas filas probablemente no son errores de registro al azar, sino reservas distintas** (por ejemplo, varias habitaciones de un mismo grupo registradas por separado), aunque sin un identificador de reserva no se puede confirmar:

- **No se distribuyen como el resto de los datos**. Un error de captura (doble envío, carga repetida) afectaría a cualquier tipo de reserva por igual, pero aquí los duplicados se concentran en ciertos perfiles. El segmento `Groups` representa el **41.6 %** de las filas duplicadas y solo el **16.6 %** del total; `Offline TA/TO` pasa de 20.3 % a 30.4 %. En cambio, `Online TA`, que es el 47.3 % del total, solo es el 20.6 % de los duplicados. El depósito `Non Refund` sube de **12.2 %** a **35.6 %**, y la tasa de cancelación de las filas duplicadas es **0.584** frente a **0.370** del total. Además, el *City Hotel* tiene más del doble de proporción de sobrantes que el *Resort Hotel* (32.65 % vs 15.21 %).
- **Los grupos son demasiado grandes para ser errores**. Aunque 5,034 grupos son pares, hay 97 grupos de 20 copias, y el grupo más grande tiene **180 filas idénticas**: *City Hotel*, segmento `Groups`, `Non Refund`, cancelada, llegada el 7 de noviembre de 2016, reservada con 277 días de anticipación y cancelada el 2016-04-04. Es difícil que un formulario se envíe 180 veces por error; es más creíble que una agencia bloqueó 180 habitaciones y canceló el bloque completo el mismo día.
- La vista de 2.1 muestra que también hay grupos pequeños (pares y tríos) de reservas `Direct`, `Online TA` o `Corporate` que terminaron en `Check-Out`. En esos casos un doble registro es más plausible, aunque también podrían ser, por ejemplo, dos parejas que reservaron la misma estancia con la misma tarifa.

Por eso, cada fila repetida puede ser una **habitación o reserva real** y no una copia sobrante. Probablemente el dataset mezcla ambos casos, y sin un identificador no hay forma de separar las copias erróneas de las reservas legítimas que coinciden en todo.

**b)** `duplicated()` trata todos los `NaN` como **iguales entre sí** (a diferencia de `==`, donde `np.nan == np.nan` es `False`). Dos reservas sin `agent` o sin `company` se consideran iguales en esa columna. Esto importa porque `company` está vacía en casi todo el dataset, y el grupo más grande de 180 filas tiene `NaN` tanto en `agent` como en `company`. Un faltante no significa "la misma agencia o empresa", sino "no sabemos cuál". Reservas hechas por entidades distintas pero no registradas terminan viéndose idénticas, y eso **aumenta la probabilidad de marcar como duplicadas reservas diferentes**. Además, `value_counts()` con el valor por defecto `dropna=True` descarta las combinaciones con `NaN` y solo cuenta **217 filas** de 119,390. Por eso fue necesario usar `dropna=False`: sin él, el análisis del tamaño de los grupos habría ignorado casi todo el dataset.

---
## Actividad 3. Impacto en el análisis (25 puntos)

### 3.1 Versión sin duplicados (5 pts)

Crea `hoteles_limpio` eliminando los duplicados exactos, conservando la primera aparición y **renumerando el índice**. Imprime el número de filas antes y después y cuántas se eliminaron.

In [9]:
# drop_duplicates() sin subset: elimina duplicados exactos (todas las columnas);
# keep='first' conserva la primera aparición e ignore_index=True renumera el índice 0, 1, 2, ...
# Se asigna a una variable nueva para no modificar hoteles
hoteles_limpio = hoteles.drop_duplicates(keep='first', ignore_index=True)

print(f'Filas antes    : {len(hoteles):,}')
print(f'Filas después  : {len(hoteles_limpio):,}')
print(f'Filas eliminadas: {len(hoteles) - len(hoteles_limpio):,}')

Filas antes    : 119,390
Filas después  : 87,396
Filas eliminadas: 31,994


### 3.2 Métricas por hotel (8 pts)

Para `hoteles` y para `hoteles_limpio`, calcula por hotel con `groupby()`:

- Número de reservas (`.size()`).
- Tasa de cancelación (media de `is_canceled`).
- Tarifa diaria promedio (media de `adr`).

Nota: para calcular la media de varias columnas a la vez, selecciona una lista de columnas: `df.groupby('hotel')[['col1', 'col2']].mean()`.

Guarda las medias en dos `DataFrame`, `resumen_original` y `resumen_limpio`, con `hotel` como columna (usa `reset_index()`).

In [10]:
# Medias por hotel de is_canceled (tasa de cancelación) y adr (tarifa diaria promedio)
resumen_original = hoteles.groupby('hotel')[['is_canceled', 'adr']].mean()
resumen_limpio = hoteles_limpio.groupby('hotel')[['is_canceled', 'adr']].mean()

# Número de reservas por hotel con size(); se alinea por el índice (hotel)
resumen_original['reservas'] = hoteles.groupby('hotel').size()
resumen_limpio['reservas'] = hoteles_limpio.groupby('hotel').size()

# reset_index() convierte el índice hotel en una columna normal
resumen_original = resumen_original.reset_index()
resumen_limpio = resumen_limpio.reset_index()

print('Original:')
print(resumen_original)
print('\nSin duplicados:')
print(resumen_limpio)

Original:
          hotel  is_canceled         adr  reservas
0    City Hotel     0.417270  105.304465     79330
1  Resort Hotel     0.277634   94.952930     40060

Sin duplicados:
          hotel  is_canceled         adr  reservas
0    City Hotel     0.300386  110.985944     53428
1  Resort Hotel     0.234809   99.025346     33968


### 3.3 Tabla comparativa (6 pts)

Agrega a cada resumen una columna `version` (`'original'` y `'sin duplicados'`) y únelos con `pd.concat()` en un `DataFrame` llamado `comparacion`, con índice renumerado y ordenado por `hotel`, de modo que las dos versiones de cada hotel queden juntas.

In [11]:
# Columna que identifica la versión de cada resumen
resumen_original['version'] = 'original'
resumen_limpio['version'] = 'sin duplicados'

# pd.concat() apila ambos resúmenes; se ordena por hotel (y por versión, para que
# 'original' quede siempre antes que 'sin duplicados') y se renumera el índice
comparacion = (
    pd.concat([resumen_original, resumen_limpio])
    .sort_values(by=['hotel', 'version'])
    .reset_index(drop=True)
)

comparacion

,hotel,is_canceled,adr,reservas,version
0,City Hotel,0.417270,105.304465,79330,original
1,City Hotel,0.300386,110.985944,53428,sin duplicados
2,Resort Hotel,0.277634,94.952930,40060,original
3,Resort Hotel,0.234809,99.025346,33968,sin duplicados


### 3.4 Interpretación (6 pts)

a) ¿Qué métrica cambia más al eliminar duplicados? ¿En qué hotel el cambio es mayor? Explica por qué, relacionándolo con lo encontrado en la Actividad 2.

b) Si se entrenara un modelo para predecir `is_canceled` con los datos originales, ¿qué problemas causarían los duplicados en el entrenamiento y en la evaluación del modelo?

**Respuesta 3.4:**

**a)** La métrica que más cambia es el **número de reservas**: el *City Hotel* pasa de 79,330 a 53,428 (−32.65 %) y el *Resort Hotel* de 40,060 a 33,968 (−15.21 %). Entre las métricas de tasa y promedio, la que más cambia es la **tasa de cancelación**. En el *City Hotel* baja de **0.4173 a 0.3004** (−11.7 puntos porcentuales, casi un 28 % menos); en el *Resort Hotel* baja de 0.2776 a 0.2348 (−4.3 puntos). La tarifa promedio sube menos: de 105.30 a 110.99 en el *City Hotel* (+5.4 %) y de 94.95 a 99.03 en el *Resort Hotel* (+4.3 %). Como la media sube al quitarlas, las filas eliminadas tenían tarifas por debajo del promedio.

El cambio es mayor en el **City Hotel**, y esto concuerda con la Actividad 2. Ahí se concentra la mayor proporción de duplicados (32.65 % contra 15.21 %), y las filas duplicadas son sobre todo reservas canceladas (tasa de 0.584 contra 0.370 del total), `Non Refund` y del segmento `Groups`. Al eliminarlas se quitan sobre todo cancelaciones, y por eso la tasa de cancelación cae. Con los datos originales, el *City Hotel* parecía cancelar mucho más que el *Resort Hotel* (41.7 % vs 27.8 %, 14.0 puntos de diferencia). Sin duplicados la brecha se reduce a 6.6 puntos (30.0 % vs 23.5 %).

**b)** Con los datos originales, un modelo para predecir `is_canceled` tendría estos problemas:

- **Entrenamiento**: el modelo vería muchas veces el mismo ejemplo (hasta 180 veces en el grupo más grande) y le daría más peso del que le corresponde, lo que favorece el **sobreajuste**. Además, la clase positiva aparecería inflada: la tasa de cancelación pasa de 0.370 en el total a 0.584 en las filas duplicadas, así que el modelo aprendería una proporción de cancelaciones mayor que la real y patrones sesgados hacia perfiles como `Groups` + `Non Refund` = cancelada.
- **Evaluación**: al dividir en entrenamiento y prueba, una fila puede quedar en entrenamiento y su copia idéntica en prueba (**fuga de datos** o *data leakage*). El modelo "acierta" porque ya vio exactamente ese caso, y la exactitud reportada resulta **optimista**: no refleja cómo se comportaría con reservas nuevas.

---
## Actividad 4. Tratamiento y verificación (20 puntos)

### 4.1 Verificación (5 pts)

Verifica sobre `hoteles_limpio` que:

- No quedan duplicados exactos.
- No hay índices repetidos y el índice va de 0 a `len(hoteles_limpio) - 1`.

In [12]:
# 1) Duplicados exactos restantes: debe ser 0
print('Duplicados exactos restantes:', hoteles_limpio.duplicated().sum())

# 2) Índices repetidos: index.duplicated() compara el índice, no los valores; debe ser 0
print('Índices repetidos           :', hoteles_limpio.index.duplicated().sum())

# 3) El índice debe ir de 0 a len - 1 (renumerado por ignore_index=True)
print('Índice mínimo               :', hoteles_limpio.index.min())
print('Índice máximo               :', hoteles_limpio.index.max())
print('len(hoteles_limpio) - 1     :', len(hoteles_limpio) - 1)

# Resumen de la verificación: True solo si se cumplen las tres condiciones
verificacion = (
    hoteles_limpio.duplicated().sum() == 0
    and hoteles_limpio.index.duplicated().sum() == 0
    and hoteles_limpio.index.min() == 0
    and hoteles_limpio.index.max() == len(hoteles_limpio) - 1
)
print('\n¿Limpieza e índice correctos?:', verificacion)

Duplicados exactos restantes: 0
Índices repetidos           : 0
Índice mínimo               : 0
Índice máximo               : 87395
len(hoteles_limpio) - 1     : 87395

¿Limpieza e índice correctos?: True


### 4.2 Duplicados parciales (10 pts)

La columna `reservation_status_date` registra la fecha del último cambio de estado. Detecta las filas de `hoteles_limpio` que **coinciden en todas las columnas excepto `reservation_status_date`**.

1. Cuenta cuántas filas sobrantes hay con ese criterio.
2. Muestra algunas de esas filas ordenadas, con las columnas `hotel`, `lead_time`, `arrival_date_year`, `arrival_date_month`, `arrival_date_day_of_month`, `adr`, `reservation_status` y `reservation_status_date`.
3. Muestra la distribución de `reservation_status` en esas filas.

Después responde: ¿qué podrían representar estas filas? ¿Las eliminarías? Justifica.

In [13]:
# Criterio parcial: todas las columnas EXCEPTO reservation_status_date
columnas_sin_fecha = hoteles_limpio.columns.drop('reservation_status_date')

# 1) Filas sobrantes con ese criterio (keep='first') y filas involucradas (keep=False)
n_parciales = hoteles_limpio.duplicated(subset=columnas_sin_fecha).sum()
mascara_parcial = hoteles_limpio.duplicated(subset=columnas_sin_fecha, keep=False)

print(f'Filas sobrantes (duplicados parciales): {n_parciales}')
print(f'Filas involucradas                    : {mascara_parcial.sum()}')
print(f'Grupos                                : {mascara_parcial.sum() - n_parciales}')

# 2) Filas involucradas ordenadas por las columnas del criterio, para que cada grupo quede junto
columnas_parcial = ['hotel', 'lead_time', 'arrival_date_year', 'arrival_date_month',
                    'arrival_date_day_of_month', 'adr', 'reservation_status',
                    'reservation_status_date']

hoteles_limpio[mascara_parcial].sort_values(by=list(columnas_sin_fecha))[columnas_parcial].head(12)

Filas sobrantes (duplicados parciales): 245
Filas involucradas                    : 471
Grupos                                : 226


,hotel,lead_time,arrival_date_year,arrival_date_month,arrival_date_day_of_month,adr,reservation_status,reservation_status_date
35572,City Hotel,9,2015,September,10,95.00,Canceled,2015-09-03
35574,City Hotel,9,2015,September,10,95.00,Canceled,2015-09-01
51267,City Hotel,12,2017,July,15,139.00,Canceled,2017-07-10
51268,City Hotel,12,2017,July,15,139.00,Canceled,2017-07-07
54170,City Hotel,14,2015,October,15,100.00,Canceled,2015-10-07
54474,City Hotel,14,2015,October,15,100.00,Canceled,2015-10-13
39425,City Hotel,14,2016,April,16,127.67,Canceled,2016-04-12
39432,City Hotel,14,2016,April,16,127.67,Canceled,2016-04-04
40080,City Hotel,14,2016,May,8,159.00,Canceled,2016-04-24
40083,City Hotel,14,2016,May,8,159.00,Canceled,2016-04-26


In [14]:
# 3) Distribución de reservation_status en las filas involucradas en duplicados parciales
print(hoteles_limpio[mascara_parcial]['reservation_status'].value_counts())

reservation_status
Canceled    471
Name: count, dtype: int64


**Respuesta 4.2:**

Con el criterio de todas las columnas excepto `reservation_status_date` hay **245 filas sobrantes**; en total, 471 filas de `hoteles_limpio` forman parte de alguno de los 226 grupos. Las **471 filas tienen estado `Canceled`**, y en cada par las filas coinciden en hotel, anticipación, fecha de llegada, tarifa y demás columnas, y solo difieren en la fecha en que se registró la cancelación, que en los ejemplos mostrados está a entre 2 y 8 días de distancia (por ejemplo, 2015-09-01 y 2015-09-03 para la misma llegada del 10 de septiembre de 2015).

**Qué podrían representar**: hay dos explicaciones posibles. (1) Una misma reserva cuya cancelación quedó registrada dos veces en fechas distintas (un error o una actualización del sistema). (2) Dos reservas distintas e iguales, por ejemplo un cliente o agencia que hizo la misma reserva dos veces (o dos habitaciones) y las canceló en días diferentes. Que siempre sean canceladas y que la diferencia sea solo la fecha de cancelación es compatible con ambas, y sin identificador de reserva no hay forma de saber cuál es la correcta.

**¿Las eliminaría?** **No las eliminaría automáticamente**. A diferencia de los duplicados exactos, estas filas **no son iguales**: difieren en un dato real (la fecha del cambio de estado), así que es un caso de **conflicto** que requiere revisión, no un duplicado exacto. Si elimináramos una, `keep` decidiría qué fecha sobrevive según el orden de las filas, no según un criterio justificado. Además, el impacto es pequeño: 245 filas son el 0.28 % de las 87,396 que quedan, así que conservarlas no distorsiona las métricas. Lo prudente es **marcarlas y reportarlas** al proveedor de los datos para que confirme si son la misma reserva.

### 4.3 Registro de decisión (5 pts)

Redacta, en una celda de texto, el registro de la limpieza realizada, como lo documentarías en un proyecto real. Incluye:

- **Criterio** de duplicado utilizado (columnas comparadas).
- **Regla de conservación** (`keep`) y por qué es o no relevante en este caso.
- **Resultado** (filas antes, después y eliminadas).
- **Supuestos y limitaciones**.

**Respuesta 4.3:**

**Registro de limpieza: duplicados en Hotel Booking Demand**

- **Criterio de duplicado**: duplicidad **exacta**. Dos filas son duplicadas si coinciden en **las 32 columnas** (`subset=None`). Los `NaN` (por ejemplo, en `agent` y `company`) se consideran iguales entre sí, que es el comportamiento de `duplicated()`/`drop_duplicates()`.
- **Regla de conservación**: `keep='first'` (se conserva la primera aparición de cada grupo) con `ignore_index=True`. En este caso **la elección de `keep` no es relevante para el contenido**: como las copias son idénticas en todas las columnas, `'first'` y `'last'` dejan exactamente los mismos valores (ambos marcan 31,994 filas) y solo cambiaría qué índice original sobrevive. Esa diferencia también desaparece al renumerar el índice. Sí sería relevante si las copias difirieran en alguna columna, como los duplicados parciales de 4.2.
- **Resultado**: filas antes **119,390**, después **87,396**, eliminadas **31,994** (26.80 %). Por hotel se eliminaron 25,902 filas del *City Hotel* (32.65 % de sus reservas) y 6,092 del *Resort Hotel* (15.21 %). Verificación: 0 duplicados exactos, 0 índices repetidos, índice de 0 a 87,395.
- **Duplicados parciales**: 245 filas sobrantes que coinciden en todo excepto `reservation_status_date` (todas `Canceled`). **Se conservaron** y quedan marcadas para revisión.
- **Supuestos**: (1) dos filas idénticas en las 32 columnas representan la misma reserva registrada más de una vez; (2) un `NaN` en `agent` o `company` equivale a otro `NaN`.
- **Limitaciones**:
  - El dataset **no tiene identificador de reserva**, así que el supuesto (1) no se puede verificar. Los duplicados se concentran en `Groups` (41.6 % vs 16.6 %), `Non Refund` (35.6 % vs 12.2 %) y reservas canceladas (0.584 vs 0.370), y hay grupos de hasta 180 filas. Esto sugiere que parte de ellos son habitaciones distintas de un mismo grupo y no errores, por lo que la limpieza podría estar **subestimando** la demanda real.
  - La limpieza cambia las métricas: la tasa de cancelación del *City Hotel* baja de 0.417 a 0.300 y la del *Resort Hotel* de 0.278 a 0.235; la tarifa promedio sube 5.68 y 4.07 respectivamente. Cualquier reporte debe indicar qué versión de los datos usa.
  - `hoteles_limpio` es adecuado para analizar el comportamiento por reserva y para entrenar modelos (evita la fuga de datos); para análisis de ocupación o volumen de habitaciones conviene usar la versión original.

---
## Reflexión final (10 puntos)

Responde brevemente, con base en tus resultados:

1. Describe una situación en la que **no** convendría eliminar los duplicados exactos de este dataset.
2. ¿Qué información adicional (columnas) le pedirías al proveedor de los datos para tomar una decisión más segura? ¿Por qué?
3. En tus propias palabras, ¿por qué `duplicated().sum() == 0` no garantiza que un dataset esté libre de duplicados?

**Respuesta final:**

1. **Cuándo no convendría eliminarlos**: cuando el objetivo es medir **volumen u ocupación**, por ejemplo cuántas habitaciones se reservaron o cancelaron en una fecha para planear personal o ingresos. En este dataset los duplicados se concentran en el segmento `Groups` y hay grupos de hasta 180 filas idénticas; si cada fila es una habitación de un mismo bloque, eliminar las copias dejaría una sola habitación donde hubo 180. El *City Hotel* perdería el 32.65 % de sus registros y su tasa de cancelación bajaría de 41.7 % a 30.0 %, lo que ocultaría las cancelaciones masivas de grupos, que son justamente un riesgo importante para el hotel.

2. **Información adicional**: pediría (a) un **identificador único de reserva**, que resolvería directamente si dos filas son la misma reserva o reservas distintas; (b) un **identificador de grupo o bloque** y el **número de habitación**, para saber si las filas repetidas son habitaciones distintas de un mismo paquete (como el grupo de 180 filas de `Groups`); y (c) la **fecha y hora de creación** de la reserva, porque dos reservas idénticas creadas en momentos distintos serían legítimas, mientras que dos creadas en el mismo segundo apuntarían a un doble registro. Con esa información también se podrían resolver los 245 duplicados parciales de 4.2.

3. **Por qué `duplicated().sum() == 0` no basta**: solo garantiza que no quedan filas **idénticas en todas las columnas**. En la Actividad 4.1 `hoteles_limpio` da 0 duplicados exactos y, sin embargo, en 4.2 aparecen **245 filas** que coinciden en todo excepto `reservation_status_date`: basta con que una sola columna difiera para que `duplicated()` no las detecte. Lo mismo pasa con duplicados ocultos por formato (`"PRT"` vs `"prt "`), por tipos de dato distintos (`101` vs `"101"`) o por identificadores diferentes asignados a un mismo registro. Además, `duplicated()` depende del criterio: con `subset=None` responde "¿hay filas idénticas?", no "¿hay registros que representan la misma reserva?". Esa segunda pregunta requiere definir un criterio de duplicado y revisar los datos.

---
**Fin de la actividad.** Antes de entregar: **Reiniciar y ejecutar todo**, verificar que no haya errores, guardar con las salidas visibles y comprobar que el enlace abre sin iniciar sesión.